custom/enikshay/tasks.py

```diff
@@ -638,3 +582,63 @@ def _get_diagnostic_test_name(self, diagnostic_test):
                 diagnostic_test.get_case_property('investigation_type_name'), site_specimen_name)
         else:
             return diagnostic_test.get_case_property('investigation_type_name')
+
+
+def calculate_dose_status_by_day(adherence_cases):
+    """
+    adherence_cases: list of 'adherence' case dicts that come from elasticsearch
+    Returns: {day: DoseStatus(taken, missed, unknown, source)}
+    """
+
+    adherence_cases_by_date = defaultdict(list)
+    for case in adherence_cases:
+        adherence_date = parse_date(case['adherence_date']) or parse_datetime(case['adherence_date']).date()
+        adherence_cases_by_date[adherence_date].append(case)
+
+    status_by_day = defaultdict(lambda: DoseStatus(taken=False, missed=False, unknown=True, source=False))
+    for d, cases in adherence_cases_by_date.iteritems():
+        case = _get_relevent_case(cases)
+        if not case:
+            pass  # unknown
+        elif case.get('adherence_value') in DOSE_TAKEN_INDICATORS:
+            source = case.get('adherence_report_source') or case.get('adherence_source')
+            status_by_day[d] = DoseStatus(taken=True, missed=False, unknown=False, source=source)
+        elif case.get('adherence_value') == DOSE_MISSED:
+            status_by_day[d] = DoseStatus(taken=False, missed=True, unknown=False, source=False)
+        else:
+            pass  # unknown
+    return status_by_day
+
+
+def _get_relevent_case(cases):
+    """
+    If there are multiple cases on one day filter to a single case as per below
+    1. Find most relevant case
+        if only non-enikshay source cases
+            consider the case with latest_modified - irrespective of case is closed/open
+
+        if only enikshay source cases
+            filter by '(closed and closure_reason == HISTORICAL_CLOSURE_REASON) or open'
+            consider the case with latest modified after above filter
+
+        if mix of enikshay and non-enikshay source cases
+            ignore non-enikshay and apply above enikshay only condition
+    2. Check if 'adherence_value' of most relevent case is one of DOSE_TAKEN_INDICATORS
+    """
+    sources = set(map(lambda x: x["adherence_source"], cases))
+    if 'enikshay' not in sources:
+        valid_cases = cases
+    else:
+        valid_cases = filter(
+            lambda case: (
+                case.get('adherence_source') == 'enikshay' and
+                (not case['closed'] or (case['closed'] and
+                    case.get('adherence_closure_reason') == HISTORICAL_CLOSURE_REASON))
+            ),
+            cases
+        )
+    if valid_cases:
+        by_modified_on = sorted(valid_cases, key=lambda case: case['modified_on'])
+        latest_case = by_modified_on[-1]
+        return latest_case
+    return None
```

Comment:
Do we need to return `None`? return would anyway be `None` unless doing it for clarity.